### ISSUE TITLE
[API Refactor] - Would love to help

### ISSUE THREAD

**Issue by lacoperon** [williamgilpin/pypdb#17]
Hi there,

I'd love to help with the refactor to use the new RPDB API to keep this package up-to-date and useable.

Are there any functions / etc you'd like me to start with?

Cheers,
- lacoperon

**Comment by williamgilpin** [comment]
Hi @lacoperon I would very much appreciate any help. I just pushed a really dirty update to the GitHub version, which fixes the basic search and retrieval functions: `Query("text").search()`, `get_all_info()`, `get_info()`

But I could really use some help:
+  Migrating old functions over, keeping the input/output as similar as possible to avoid breaking existing applications. Most of the functions from the original version need to be migrated, although it's not clear to me whether people use many of them (so they might be low priority). 
+ Making this package work much more elegantly, since the project structure is somewhat hacked together, especially now. There is a lot that the new RCSB API can do, and so there might be some interesting core work to be done converting the `Query()` object into something more flexible (especially since, currently, adding new query types requires hard-coding them in, which seems gratuitous).

Some links I found helpful today are [the new API reference](https://www.rcsb.org/pages/webservices) and the tiny "JSON" link that appears whenever one does an advanced search on the website, [like this one](https://www.rcsb.org/search?request=%7B%22query%22%3A%7B%22type%22%3A%22group%22%2C%22logical_operator%22%3A%22and%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22group%22%2C%22logical_operator%22%3A%22and%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22group%22%2C%22logical_operator%22%3A%22and%22%2C%22nodes%22%3A%5B%7B%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22parameters%22%3A%7B%22operator%22%3A%22contains_phrase%22%2C%22negation%22%3Afalse%2C%22value%22%3A%22test%22%2C%22attribute%22%3A%22struct.title%22%7D%2C%22node_id%22%3A0%7D%5D%7D%5D%2C%22label%22%3A%22text%22%7D%5D%2C%22label%22%3A%22query-builder%22%7D%2C%22return_type%22%3A%22entry%22%2C%22request_options%22%3A%7B%22pager%22%3A%7B%22start%22%3A0%2C%22rows%22%3A100%7D%2C%22scoring_strategy%22%3A%22combined%22%2C%22sort%22%3A%5B%7B%22sort_by%22%3A%22score%22%2C%22direction%22%3A%22desc%22%7D%5D%7D%2C%22request_info%22%3A%7B%22src%22%3A%22ui%22%2C%22query_id%22%3A%220760a2a2a12eb198e8f63f63926310dd%22%7D%7D)

**Comment by lacoperon** [comment]
Sweet, sounds good to me. I just want to confirm a couple of design principles I was hoping to confirm with you, to make sure that my implementation won't hinder your groove in future. 

Note that these are scoped at the bare bones API level. For example, this does not include:
* searching for structures without ligands
* searching for related structures
* searching for structures by experimental method

## Design Thoughts

### 1. Refactoring into multiple files

I'm thinking I might create separate files for each API service.
For example, logic for the entity service could be within `entity_service.py` or some such.

The main `pypdb.py` file (in my mind) might contain nothing but query functions for each endpoint service, pointing to the separate file (e.g. `clients/entity_service.py`)

I think this would be nice, because then the PyPDB code structure will be consistent with the API, and readable at a glance within short-ish files.

### 2. Adding dataclasses for API Requests/Responses

I was hoping to explicitly spell out what requests and responses from each service endpoint look like, by defining `$ENDPOINTRequest` and `$ENDPOINTResponse` objects using [dataclasses](https://docs.python.org/3/library/dataclasses.html) for each service in the API.

The idea would be that the dataclass definitions exactly match the RSCB API docs. These objects would also have the API ReDoc linked to them in their docstrings, so it's trivial for contributors to add fields that might be added to the API in future.

For example, `EntityServiceRequest` might look like:
```python
ENTITY_SERVICE_BASE_URL = "https://data.rcsb.org/rest/v1/core/polymer_entity/{entry_id}/{entity_id}"

@dataclass 
class EntityServiceRequest:
"""For documentation on these request fields, see: 
https://data.rcsb.org/redoc/index.html#tag/Entity-Service
"""
  entry_id: str
  entity_id: str

  def to_url(self) -> str:
    return ENTITY_SERVICE_BASE_URL.format(entry_id,  entity_id)
```
`EntityServiceResponse` would look way more complicated, but that complicated definition would aim to mirror the definition in the API. If more fields are added to the EntityService response in the API itself, it would be trivial to add the field to the wrapper.

For example:
```python
from typing import List
from somewhere import EntityPoly, RcsbClusterMembership

class EntityServiceResponse:
"""For documentation on these response fields, see:
https://data.rcsb.org/redoc/index.html#tag/Entity-Service
"""
  rcsb_cluster_membership : List[RcsbClusterMembership] 
  entity_poly : EntityPoly
  rcsb_id : str
  # etc etc etc

  @classmethod
  def from_json(cls, json_object : JsonDumpType) -> EntityServiceResponse:
     """This is a constructor function to make `EntityServiceResponse` instances from RCSB API JSON responses."""
    rcsb_cluster_membership = [RcsbClusterMembership.from_json(cluster_member_json) 
      for cluster_member_json in json_object.rcsb_cluster_membership]

    return cls(entity_poly=EntityPoly.from_json(json_object.entity_poly),
      rcsb_cluster_membership=rcsb_cluster_membership,
      rcsb_id=json_object.rcsb_id)
```
(we could recursively define logic within other classes to parse RcsbClusterMembership, EntityPoly etc from JSON dump object).
                       
### 3. Adding type annotations / enforcement using `mypy`

Supposing we agree on the dataclass approach in #2, we kind of get this for 'free'. [PEP 526](https://www.python.org/dev/peps/pep-0526/) added type annotations to Python, and [`mypy`](https://mypy.readthedocs.io/en/stable/existing_code.html) can enforce that within software packages.

The benefit of this is that we will catch any breaking API changes that might affect upstream functions by getting lint warnings.

### What this would look like within `pypdb.py`.

For example, within `pypdb.py`, we could have one of these types of functions for each endpoint, with examples.
(I omit the examples here).
```python
from clients import entity_service
from somewhere import url_to_json_response
def query_entity_service(entry_id : str, entity_id : str) -> entity_service.EntityServiceResponse:
  query_url = entity_service.EntityServiceRequest(entry_id, entity_id).to_url()
  json_dump = url_to_json_response(query_url)
  return entity_service.EntityServiceResponse.from_json(json_dump)
```

**TLDR** LMK if that all sounds good to you.

**Comment by williamgilpin** [comment]
That all sounds great to me, I would love for this refactor to be an opportunity to make this repo more scalably structured and broadly useful. The only thing to keep in mind is to the call structure intact for a few widely-used legacy functions (like `get_all_info` and keyword searches), although we can always add deprecation warnings---I get the impression that those particular functions are used a lot downstream in other apps. However, since the output format of `get_all_info` and the other data retrieval functions changed with the new API, a lot of downstream apps are probably already broken---so it might not matter.

If you want to do a PR with a few key functions implemented, then once I go through it and figure out the structure, I can start creating more functions in the same style.

Just a heads up that I can be a bit sporadic with when I have time to work on this, but I very much am excited about improving it (and hopefully learning some new stuff along the way)

**Comment by lacoperon** [comment]
Closed given I think we're on the same page.
